### PR TITLE
Adding Machine Learning Compute cmdlets

### PR DESCRIPTION
**Pull request by shutchings**
## Description
This is an approved PR for the 4.5.0 release. These cmdlet's were supposed to make it in to the 4.4.0 release but didn't quite make it. I've been working closely with @markcowl and @cormacpayne for my reviews and feedback. [This was the original PR](https://github.com/Azure/azure-powershell-pr/pull/292). A lot of feedback was incorporated into it but @cormacpayne closed it in favor of this [PR](https://github.com/Azure/azure-powershell-pr/pull/356) where he continued to merge some of the recent changes from preview.

The .NET SDK was merged to the public repo in this [PR](https://github.com/Azure/azure-sdk-for-net/pull/3723). The nuget package has been published and these cmdlets are **not** using a local package.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by shutchings** [comment]
Looks like @cormacpayne and @markcowl are OOF until the end of the week. Is someone stepping in for them?

**Comment by cormacpayne** [comment]
@azuresdkci add to whitelist

File: ChangeLog.md [hunk 1]
```diff
@@ -1,4 +1,13 @@
-﻿## 2017.09.25 - Version 4.4.0
```
Comment (cormacpayne) [inline ChangeLog.md]: @shutchings please revert the changes made to this file, we will handle updating the ChangeLog in the root of the repository during the release

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCOmpute/app.config [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCOmpute/app.config]: @shutchings there should be no `app.config` file for the `Commands` folder

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Commands.MachineLearningCompute.csproj [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      <DesignTime>True</DesignTime>
+      <DependentUpon>Resources.resx</DependentUpon>
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="..\AzureRM.MachineLearningCompute.psd1">
+      <Link>AzureRM.MachineLearningCompute.psd1</Link>
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
+    <None Include="app.config" />
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Commands.MachineLearningCompute.csproj]: @shutchings please remove the reference to `app.config`

File: ChangeLog.md (hunk 1 above)
Comment (shutchings) [reply ChangeLog.md]: Sounds good!

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCOmpute/app.config (hunk 2 above)
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCOmpute/app.config]: Got it. I think I had it because of some dependency issues but it looks like those got cleaned up in past revisions. Probably from the changes you made.

**Comment by shutchings** [comment]
@cormacpayne I installed the package and tested the cmdlets and things look good.

**Comment by shutchings** [comment]
@cormacpayne @markcowl Looks like today is when it's supposed to be code complete. Any additional changes?

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+using Microsoft.Azure.Management.MachineLearningCompute;
+using Microsoft.Azure.Management.MachineLearningCompute.Models;
+using Microsoft.Rest.Azure;
+using System;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.MachineLearningCompute.Cmdlets
+{
+    [Cmdlet(VerbsCommon.Get, CmdletSuffix + "Key")]
+    [OutputType(typeof(OperationalizationClusterCredentials))]
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs]: @shutchings can we create a new class that is `PSOperationalizationClusterCredentials` and return that instead? This will prevent issues with potential breaking changes in the future.

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            else if (string.Equals(this.ParameterSetName, ResourceIdParameterSet, StringComparison.OrdinalIgnoreCase))
+            {
+                var resourceInfo = new ResourceIdentifier(ResourceId);
+                ResourceGroupName = resourceInfo.ResourceGroupName;
+                Name = resourceInfo.ResourceName;
+            }
+
+            try
+            {
+                WriteObject(MachineLearningComputeManagementClient.OperationalizationClusters.ListKeys(ResourceGroupName, Name));
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs]: @shutchings if this is returning more than one object, make sure to  use the `WriteObject` overload with `enumerateCollection:true`

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/MachineLearningComputeCmdletBase.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                        AzureSession.Instance.ClientFactory.CreateArmClient<MachineLearningComputeManagementClient>(
+                            DefaultProfile.DefaultContext, Common.Authentication.Abstractions.AzureEnvironment.Endpoint.ResourceManager));
+            }
+        }
+
+        protected void HandleNestedExceptionMessages(CloudException e)
+        {
+            var exceptionDetails = e.Message + ".";
+
+            foreach (var err in e.Body.Details)
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/MachineLearningComputeCmdletBase.cs]: @shutchings is the body ever null for an error? I want to make sure we avoid the scenario where `e.Body.Details` throws an error because `e.Body` is null.

Also, would we need to do any null-checking for `e.Body.Details`?

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/TestAzureRmMlOpClusterSystemServicesUpdateAvailability.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+using Microsoft.Azure.Management.MachineLearningCompute;
+using Microsoft.Azure.Management.MachineLearningCompute.Models;
+using Microsoft.Rest.Azure;
+using System;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.MachineLearningCompute.Cmdlets
+{
+    [Cmdlet(VerbsDiagnostic.Test, CmdletSuffix + "SystemServicesUpdateAvailability")]
+    [OutputType(typeof(CheckSystemServicesUpdatesAvailableResponse))]
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/TestAzureRmMlOpClusterSystemServicesUpdateAvailability.cs]: @shutchings same comment about wrapping this in a PS class

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/UpdateAzureRmMlOpClusterSystemService.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+using Microsoft.Azure.Management.MachineLearningCompute.Models;
+using Microsoft.Rest.Azure;
+using System;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.MachineLearningCompute.Cmdlets
+{
+
+    [Cmdlet(VerbsData.Update, CmdletSuffix + "SystemService", SupportsShouldProcess = true)]
+    [OutputType(typeof(UpdateSystemServicesResponse))]
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/UpdateAzureRmMlOpClusterSystemService.cs]: @shutchings same comment about wrapping this in a PS class

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/packages.config [hunk 9]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.Azure.Management.MachineLearningCompute" version="0.1.0" targetFramework="net452" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.10" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/packages.config]: @shutchings remove same packages outlined in csproj

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Commands.MachineLearningCompute.csproj [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      <Name>Commands.Common</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\..\Common\Commands.ResourceManager.Common\Commands.ResourceManager.Common.csproj">
+      <Project>{3819d8a7-c62c-4c47-8ddd-0332d9ce1252}</Project>
+      <Name>Commands.ResourceManager.Common</Name>
+    </ProjectReference>
+    <Reference Include="Microsoft.Azure.Management.MachineLearningCompute, Version=0.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.MachineLearningCompute.0.1.0\lib\net452\Microsoft.Azure.Management.MachineLearningCompute.dll</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL" />
```
[the remark this answers is not in the document]
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Commands.MachineLearningCompute.csproj]: Got it.

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs (hunk 5 above)
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs]: This method only ever returns one object that wraps up multiple keys.

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/MachineLearningComputeCmdletBase.cs (hunk 6 above)
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/MachineLearningComputeCmdletBase.cs]: Good points! I have to do this extra handling for the different types of errors that the SDK returns so that the error message to the user isn't just "Bad Request" or "Not Found". I've settled on trying to unwrap the errors as best as I can but I'll throw the original error if I have issues doing that. That way the user will always still get the original error, or a more descriptive one, and not an error due to the unwrapping.

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs (hunk 4 above)
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/GetAzureRmMlOpClusterKey.cs]: Done.

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/TestAzureRmMlOpClusterSystemServicesUpdateAvailability.cs (hunk 7 above)
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/TestAzureRmMlOpClusterSystemServicesUpdateAvailability.cs]: Done.

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/UpdateAzureRmMlOpClusterSystemService.cs (hunk 8 above)
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/UpdateAzureRmMlOpClusterSystemService.cs]: Done.

File: src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/packages.config (hunk 9 above)
Comment (shutchings) [reply src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/packages.config]: Done.

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/job/powershell-demand/1899
Packaging job: http://azuresdkci.cloudapp.net/job/posh-pack/21/

**Comment by shutchings** [comment]
@cormac The cmdlets in the package job look good. The help looks like it's out of date though. I updated the help after changing to the PS output objects. What has to be done to get the help updated in the installer? Do I have to re-update the installer?
